Mathlib/Algebra/Hom/Equiv/Basic.lean

```diff
@@ -0,0 +1,659 @@
+  ⟨fun f =>
+    { toFun := f, invFun := EquivLike.inv f, left_inv := EquivLike.left_inv f, right_inv := EquivLike.right_inv f,
+      map_mul' := MulEquivClass.map_mul f }⟩
+
+namespace MulEquiv
+
+@[to_additive]
+instance [Mul M] [Mul N] : CoeFun (M ≃* N) fun _ => M → N := ⟨fun f => f.toFun⟩
+
+@[to_additive]
+instance [Mul M] [Mul N] : MulEquivClass (M ≃* N) M N where
+  coe f := f.toFun
+  inv f := f.invFun
+  left_inv f := f.left_inv
+  right_inv f := f.right_inv
+  coe_injective' f g h₁ h₂ := by
+    cases f
+    cases g
+    congr
+    apply Equiv.coe_fn_injective h₁
+  map_mul := map_mul'
+
+variable [Mul M] [Mul N] [Mul P] [Mul Q]
+
+@[simp, to_additive]
+theorem to_equiv_eq_coe (f : M ≃* N) : f.toEquiv = f :=
+  rfl
+#align mul_equiv.to_equiv_eq_coe MulEquiv.to_equiv_eq_coe
+
+@[simp, to_additive]
+theorem to_fun_eq_coe {f : M ≃* N} : f.toFun = f :=
+  rfl
+#align mul_equiv.to_fun_eq_coe MulEquiv.to_fun_eq_coe
+
+@[simp, to_additive]
+theorem coe_to_equiv {f : M ≃* N} : (f : M ≃ N) = f.toFun :=
+  rfl
+#align mul_equiv.coe_to_equiv MulEquiv.coe_to_equiv
+
+-- Porting note: restore `to_additive`
+@[simp]
+theorem coe_toMulHom {f : M ≃* N} : f.toMulHom = f.toFun :=
+  rfl
+#align mul_equiv.coe_to_mul_hom MulEquiv.coe_toMulHom
+@[simp]
+theorem _root_.AddEquiv.coe_toAddHom [Add M] [Add N] {f : M ≃+ N} : f.toAddHom = f.toFun :=
+  rfl
+#align add_equiv.coe_to_add_hom AddEquiv.coe_toAddHom
+
+/-- A multiplicative isomorphism preserves multiplication. -/
+@[to_additive "An additive isomorphism preserves addition."]
+protected theorem map_mul (f : M ≃* N) : ∀ x y, f (x * y) = f x * f y :=
+  _root_.map_mul f
+#align mul_equiv.map_mul MulEquiv.map_mul
+
+/-- Makes a multiplicative isomorphism from a bijection which preserves multiplication. -/
+@[to_additive "Makes an additive isomorphism from a bijection which preserves addition."]
+def mk' (f : M ≃ N) (h : ∀ x y, f (x * y) = f x * f y) : M ≃* N :=
+  ⟨f.1, f.2, f.3, f.4, h⟩
+#align mul_equiv.mk' MulEquiv.mk'
+
+@[to_additive]
+protected theorem bijective (e : M ≃* N) : Function.Bijective e :=
+  EquivLike.bijective e
+#align mul_equiv.bijective MulEquiv.bijective
+
+@[to_additive]
+protected theorem injective (e : M ≃* N) : Function.Injective e :=
+  EquivLike.injective e
+#align mul_equiv.injective MulEquiv.injective
+
+@[to_additive]
+protected theorem surjective (e : M ≃* N) : Function.Surjective e :=
+  EquivLike.surjective e
+#align mul_equiv.surjective MulEquiv.surjective
+
+/-- The identity map is a multiplicative isomorphism. -/
+@[refl, to_additive "The identity map is an additive isomorphism."]
+def refl (M : Type _) [Mul M] : M ≃* M :=
+  { Equiv.refl _ with map_mul' := fun _ _ => rfl }
+#align mul_equiv.refl MulEquiv.refl
+
+@[to_additive]
+instance : Inhabited (M ≃* M) :=
+  ⟨refl M⟩
+
+/-- The inverse of an isomorphism is an isomorphism. -/
+@[symm, to_additive "The inverse of an isomorphism is an isomorphism."]
```

Comment:
Bug with `symm` it seems [#855](https://github.com/leanprover-community/mathlib4/pull/855)